### PR TITLE
chore: TS refactor

### PR DESCRIPTION
**Pull request by imorland**
**Fixes #0000**

**Changes proposed in this pull request:**

**Reviewers should focus on:**
Please give this a spin over locally. I did the same, but just in case I missed something

**Screenshot**

**Confirmed**

- [ ] Frontend changes: tested on a local Flarum installation.
- [ ] Backend changes: tests are green (run `composer test`).

**Required changes:**

- [ ] Related [Flarum core extension PR's](https://github.com/flarum/core/pulls): (Omit this section if irrelevant)

### CODE REVIEW

File: js/src/forum/extenders/extendDiscussionSearchSource.tsx [hunk 1]
```diff
@@ -3,7 +3,7 @@ import { extend } from 'flarum/common/extend';
 import DiscussionsSearchSource from 'flarum/forum/components/DiscussionsSearchSource';
 
 export default function extendDiscussionsSearchSource() {
-  extend(DiscussionsSearchSource.prototype, 'queryMutators', function (mutators: string[]) {
+  extend(DiscussionsSearchSource.prototype, 'queryMutators', function (mutators) {
```
Comment (DavideIadeluca) [inline js/src/forum/extenders/extendDiscussionSearchSource.tsx:6]:
```suggestion
  @ts-ignore
  extend(DiscussionsSearchSource.prototype, 'queryMutators', function (mutators: string[]) {
```

Let's still cast it as `string[]` like it was before so `mutators.push('-is:solved');`  doesn't error and for now add a `@ts-ignore`. Seems like a slight change in core would be needed

File: js/src/forum/extenders/extendDiscussionListState.tsx [hunk 2]
```diff
@@ -8,7 +8,11 @@ export default function extendDiscussionListState() {
       const negate = app.discussions.bestAnswer === '2';
       const prepend = negate ? '-' : '';
 
-      params.filter[`${prepend}solved-discussions`] = true;
+      if (!params.filter) {
+        params.filter = {};
+      }
+
+      (params.filter as Record<string, string>)[`${prepend}solved-discussions`] = 'true';
```
Comment (DavideIadeluca) [inline js/src/forum/extenders/extendDiscussionListState.tsx:15]: If we import the `PaginatedListParams` type and use that for the `params` parameter, we can save ourselves this gymnastic and leave it like it was before

```tsx
...
import type { PaginatedListParams } from 'flarum/common/states/PaginatedListState';

export default function extendDiscussionListState() {
  extend(DiscussionListState.prototype, 'requestParams', function (params: PaginatedListParams) {
```

File: js/src/forum/extenders/extendDiscussionComposer.tsx [hunk 3]
```diff
@@ -1,16 +1,18 @@
 import { extend } from 'flarum/common/extend';
 import app from 'flarum/forum/app';
 import DiscussionComposer from 'flarum/forum/components/DiscussionComposer';
+import Tag from 'flarum/tags/common/models/Tag';
```
Comment (DavideIadeluca) [inline js/src/forum/extenders/extendDiscussionComposer.tsx:4]:
```suggestion
import type Tag from 'flarum/tags/common/models/Tag';
```

File: js/src/forum/extend.ts [hunk 4]
```diff
@@ -4,7 +4,6 @@ import Extend from 'flarum/common/extenders';
 import Post from 'flarum/common/models/Post';
```
Comment (DavideIadeluca) [inline js/src/forum/extend.ts:4]:
```suggestion
import type Post from 'flarum/common/models/Post';
```

File: js/src/forum/components/SolutionSearchItem.tsx [hunk 5]
```diff
@@ -5,24 +5,24 @@ import highlight from 'flarum/common/helpers/highlight';
 import Discussion from 'flarum/common/models/Discussion';
```
Comment (DavideIadeluca) [inline js/src/forum/components/SolutionSearchItem.tsx:5]:
```suggestion
import type Discussion from 'flarum/common/models/Discussion';
```

File: js/src/forum/components/SolutionSearchItem.tsx [hunk 6]
```diff
@@ -5,24 +5,24 @@ import highlight from 'flarum/common/helpers/highlight';
...
 import Post from 'flarum/common/models/Post';
```
Comment (DavideIadeluca) [inline js/src/forum/components/SolutionSearchItem.tsx:6]:
```suggestion
import type Post from 'flarum/common/models/Post';
```

File: js/src/forum/components/SolutionSearchItem.tsx [hunk 7]
```diff
@@ -5,24 +5,24 @@ import highlight from 'flarum/common/helpers/highlight';
...
 import type Mithril from 'mithril';
-/** @ts-ignore */
-import tagsLabel from 'flarum/tags/helpers/tagsLabel';
+import tagsLabel from 'flarum/tags/common/helpers/tagsLabel';
 import ItemList from 'flarum/common/utils/ItemList';
+import Tag from 'flarum/tags/common/models/Tag';
```
Comment (DavideIadeluca) [inline js/src/forum/components/SolutionSearchItem.tsx:10]:
```suggestion
import type Tag from 'flarum/tags/common/models/Tag';
```

File: js/src/forum/components/SelectBestAnswerNotification.ts [hunk 8]
```diff
@@ -1,3 +1,4 @@
+import Discussion from 'flarum/common/models/Discussion';
```
Comment (DavideIadeluca) [inline js/src/forum/components/SelectBestAnswerNotification.ts:1]:
```suggestion
import type Discussion from 'flarum/common/models/Discussion';
```

**Review by DavideIadeluca — CHANGES_REQUESTED** [review]
Everything appears to be working locally as expected. There are some suggestions I have for typings though, including some changes I picked up that are unrelated to this PR

File: js/src/@types/shims.d.ts [hunk 9]
```diff
@@ -1,18 +1,34 @@
 import Post from 'flarum/common/models/Post';
```
[the remark this answers is not in the document]
Comment (imorland) [reply js/src/@types/shims.d.ts]: Not sure I follow here??

File: js/src/forum/components/BestAnswerBadge.tsx [hunk 10]
```diff
@@ -1,10 +1,11 @@
 import Badge, { IBadgeAttrs } from 'flarum/common/components/Badge';
 import app from 'flarum/forum/app';
+import extractText from 'flarum/common/utils/extractText';
 
 export default class BestAnswerBadge extends Badge {
   static initAttrs(attrs: IBadgeAttrs) {
     attrs.type = 'bestAnswer';
     attrs.icon = 'fas fa-check';
-    attrs.label = app.translator.trans('fof-best-answer.forum.answered_badge') as string;
+    attrs.label = extractText(app.translator.trans('fof-best-answer.forum.answered_badge'));
```
[the remark this answers is not in the document]
Comment (imorland) [reply js/src/forum/components/BestAnswerBadge.tsx:9]: I totally understand where you're coming from here, but I think this is the way forward, at least for the remaining time on `1.x`. 

Let's alter core for `2.x` as you suggest for the typing...

[the remark this answers is not in the document]
Comment (DavideIadeluca) [reply js/src/forum/components/BestAnswerBadge.tsx:9]: Sounds good to me

File: js/src/@types/shims.d.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (DavideIadeluca) [reply js/src/@types/shims.d.ts]: Sorry I could've been more clear. Check out https://github.com/flarum/framework/pull/4097/files#diff-adc7130f3dfa4774617fac483805a1b4206ba70bd0a9f86c073546e1cf36b472R1-R6 for that I mean

**Review by DavideIadeluca — APPROVED** [review]

File: js/src/@types/shims.d.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (imorland) [reply js/src/@types/shims.d.ts]: Got it, thanks!

**Review by DavideIadeluca — APPROVED** [review]

**Pull request merged**